src/syscalls.rs

```diff
@@ -51,12 +56,6 @@ pub fn yieldk() {
     }
 }
 
-pub fn yieldk_for<F: Fn() -> bool>(cond: F) {
-    while !cond() {
-        yieldk();
-    }
-}
-
```

Comment:
Removing the previous interface is a breaking change. Please don't do so without bumping the version number.

I'm not against adding an async/await API, but we need to leave time for projects that depend on libtock-rs to transition before removing the previous API. Especially given the performance concerns that have been raised on this issue.

Reply:
What are the versioning and stability guarantees? The library isn't even on crates.io, so what does a version number even mean?

I would like to remove the `yieldk_for` function because yielding makes it almost impossible for `libtock-rs` to become a safe and ergonomic library. For example, yielding in callbacks breaks the guarantees of `FnMut` closures and it also can clog up the event queue which leads to kernel panics.

If you cannot find an abstraction that is safe and ergonomic, using Rust becomes pointless. That's more or less what I am worried about.

Reply:
> What are the versioning and stability guarantees? The library isn't even on crates.io, so what does a version number even mean?

I didn't notice that it's not on crates.io. I think it would be useful there. For example, https://github.com/tock/tock/tree/master/libraries/tock-register-interface is already on crates.io. But that's a separate debate.

In any case, one can depend on this repository as a git dependency. In that case, versioning can be done by a specific commit. Still, breaking any external user of this library in one commit without much notice or tracking issue wouldn't be user-friendly. This is forcing external users to instantly migrate to the new future-based API or to fork the project if they need time to migrate but still want to benefit from the next commit that will appear here for example.

At least, bumping the version shows that the API changes, and makes it explicit that if users must now move on to the future-based API if they want to benefit from further additions to libtock-rs.

Additionally, the new API may have bugs that will take some extra testing time to find (for example due to the fact that the examples are rather simple so far), so instantly deprecating the previous API for a potentially unstable one is also problematic.

> I would like to remove the `yieldk_for` function because yielding makes it almost impossible for `libtock-rs` to become a safe and ergonomic library. For example, yielding in callbacks breaks the guarantees of `FnMut` closures and it also can clog up the event queue which leads to kernel panics.

I agree with your safety concerns, but then `yieldk` and `yieldk_for` should be marked as `unsafe` (as they should always have been) instead of outright removed now.

> If you cannot find an abstraction that is safe and ergonomic, using Rust becomes pointless. That's more or less what I am worried about.

It's not that black and white, Rust has a middle ground with `unsafe` code, which is used pervasively through Tock. I agree that reducing the amount of unsafe code is welcome whenever possible, but I disagree that having only part of a code base as safe Rust is pointless. Rust brings a lot of advantages in the safe part compared to C.

Reply:
So, what would you propose as a practical solution? I am asking because we still plan *a lot* of breaking changes. A few examples:

- Not all drivers are checked for availability => There should be an `init()` for every driver
- The drivers can be used at any position in the code => There must be a global `Runtime` struct which controls sharing of drivers
- There are `unwraps()` all around the code base => All drivers should return a meaningful result
- The result should contain information about the failed driver => All syscalls must return a `TockError` or something else
- The callbacks are hard to use. They must not be moved, so they need to be passed in from outside by `&mut` => They should not be used. Futures are better and don't have ownership problems
- All drivers work in a different way => They should work in the same way

And so on. `libtock` currently is a collection of small experiments. Stability would be a great. But to achieve this goal, I can only encourage everyone to put feedback on the design document. Then, we can implement the design and, after that, we can make a commitment on stability.

Reply:
Replied in https://github.com/tock/libtock-rs/issues/105#issuecomment-554144655

Reply:
Actually, I like the idea of incrementing the version. I also added a `CHANGELOG.md`.

Reply:
Thanks for taking the versioning concern into account, and for driving this future-based runtime!